File: algorithms/Sum Root to Leaf Numbers/Sum Root to Leaf Numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    return dfs(root)
};

const dfs = (root, path = '') => {
    if (!root.left && !root.right) return +(path + root.val)
    
    const left = root.left ? dfs(root.left, path + root.val) : 0
    const right = root.right ? dfs(root.right, path + root.val) : 0
    
    return left + right
}

File: algorithms/S/Sum Root to Leaf Numbers/Sum Root to Leaf Numbers.cpp
class Solution {
public:
    int ans=0;
    void dfs(TreeNode* root, string s){
        if(!root->left && !root->right){
            s+=to_string(root->val);
            ans+=stoi(s);
            return;
        }
        string o = s;
        s+=to_string(root->val);
        if(root->left) dfs(root->left,s);
        if(root->right) dfs(root->right,s);
        s=o;
        
    }
    int sumNumbers(TreeNode* root) {
        if(!root) return ans;
        string s = "";
        dfs(root,s);
        return ans;
    }
};
